cmd/manila-csi-plugin/main.go

```diff
@@ -142,6 +158,10 @@ func main() {
 	cmd.PersistentFlags().StringVar(&nodeID, "nodeid", "", "this node's ID")
 	cmd.MarkPersistentFlagRequired("nodeid")
 
+	cmd.PersistentFlags().StringVar(&nodeAZ, "nodeaz", "", "this node's availability zone")
```

Comment:
@gman0 availability zone of the node could be directly retrieved from the metadata service, is it required to have  input here?

Reply:
I was thinking this is an optional flag for use when e.g. the OpenStack cloud isn't running the metadata service.  I may be confused though.

Reply:
Or when this CSI plugin is used on worker nodes that aren't OpenStack/nova VMs

Reply:
Right, people are working on k8s on baremetal instances in OpenStack clouds ...

Reply:
ok, since its optional flag , incase user doesnt provide it , there should be provision to retrieve automatically, if there are large number of nodes in cluster, adding AZ manually for all the nodes wouldn't be feasible?

Reply:
I second what the guys have said :) !

> ok, since its optional flag , incase user doesnt provide it , there should be provision to retrieve automatically

The problem with defaulting to querying the Metadata Service is that in case it's not present (i.e.  _http://169.254.169.254/openstack/latest/meta_data.json_ is unreachable), it would first wait for a 30s timeout in [`http.Get()`](https://github.com/kubernetes/cloud-provider-openstack/blob/8c02748fab095675514861628ea57317bb36c263/pkg/csi/cinder/openstack/openstack_metadata.go#L39) and then it would return an error.

This error is ambiguous - we wouldn't be able to distinguish between:

(a) the plugin really is supposed to query the Metadata Service and the fact that `http.Get()` returned an error means there's a real problem in our cloud => the plugin should return an error and fail
(b) there's no Metadata Service in our cloud => ignore this error

By having no default value here, we're moving the responsibility of knowing about the presence of the Metadata Service to the cluster admins, which clears this ambiguity.

> adding AZ manually for all the nodes wouldn't be feasible?

What we're suggesting is that the admins would feed the output of `$(curl http://169.254.169.254/openstack/latest/meta_data.json | jq .availability_zone)` into the `--nodeaz` argument when deploying the plugin. It's mentioned in the documentation [here](https://github.com/kubernetes/cloud-provider-openstack/blob/fd6aa3342fb1d917a5fa79e02176075dc9cc5f7d/docs/using-manila-csi-plugin.md#enabling-topology-awareness)

Reply:
Oh and the docs are missing `-r` flag for `jq`, I'll need to fix this

Reply:
hmm, Im not sure distinguishing of error will be of that importance , we can log warning/error stating that unable to get AZ of node, need to set manually or so? 
In the doc anyway you are mentioning to set using metadata service , if that is accessible from node we could set it automatically only.  if for some issue , we got error on that then asking to set manually is good.
@gman0 @gouthampacha @tombarron Thoughts?

Reply:
@ramineni csi-cinder returns an error on `getAvailabilityZoneMetadataService()` failure:

https://github.com/kubernetes/cloud-provider-openstack/blob/master/pkg/csi/cinder/nodeserver.go#L458-L461

not acting on this error + having to fail a timeout doesn't seem very nice to be honest :P

The difference between Cinder and Manila in this regard is that Cinder will always need Nova so it's guaranteed to also have the metadata svc, whereas Manila doesn't.

Just to clarify, we're not asking users to set the node AZs manually, it's set automatically during the initialization of the container using that little shell script.

Reply:
@ramineni: I prefer this approach over coding the metadata service lookup inside the node server... it wouldn't really be a concern for operators in my opinion because manila-csi will be deployed/managed with some automation for the most part (helm, operators, ansible etc), so it'll be easy to configure per the example shared if such a metadata lookup is desired.

Reply:
> Oh and the docs are missing `-r` flag for `jq`, I'll need to fix this

You have it fixed in https://github.com/kubernetes/cloud-provider-openstack/commit/3d16188b74866e7719778ef5c0bace651206415a - you're missing it in https://github.com/kubernetes/cloud-provider-openstack/pull/939/files#diff-2a40081587db138d3f23b73622df5608R167 :)

Reply:
Avoiding the timeout is a big improvement for operators and they can use automation such as the supplied shell script to set the node AZs.

Reply:
@tombarron @gouthampacha @gman0 ok, got it . `nodeAZ` it is :) . Lets go with this only as all are agreement with this flag :)
@gman0 do you mean retry would be better in case of cinder-csi?

Reply:
@ramineni 

> do you mean retry would be better in case of cinder-csi?

IMHO the way csi-cinder is doing is the correct way. In case of a failure, the CO should do the retry.